File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> myMap;

        for (int i = 0; i < nums1.size(); i++) {
            myMap[nums1[i]] = i;
        }

        vector<int> res(nums1.size(), -1);

        deque<int> myQueue;

        myQueue.push_back(0);

        for (int i = 1; i < nums2.size(); i++) {
            if (nums2[i] <= nums2[myQueue.back()]) {
                myQueue.push_back(i);
            } else {
                while(!myQueue.empty() && nums2[myQueue.back()] < nums2[i] ) {
                    if (myMap.find(nums2[myQueue.back()]) != myMap.end()) {
                        res[myMap[nums2[myQueue.back()]]] = nums2[i];
                       
                    }
                     myQueue.pop_back();
                }
                myQueue.push_back(i);
            }
        }
        return res;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums_map = dict()

        for i in range(len(nums1)):
            nums_map[nums1[i]] = i

        stack = []

        res = [-1] * len(nums1)

        stack.append(0)

        for i in range(1, len(nums2)):
            if nums2[stack[-1]] > nums2[i]:
                stack.append(i)
            elif nums2[stack[-1]] < nums2[i]:
                while stack and nums2[stack[-1]] < nums2[i]:
                    idx = stack.pop()
                    if nums2[idx] in nums_map:
                        res[nums_map[nums2[idx]]] = nums2[i]
                
                stack.append(i)
        
        return res
